### PR TITLE
add file storage for sessions and npm tokens; allow to setup oidc scope in config

### PR DESCRIPTION
**Pull request by nikolaevn**
subj

### CODE REVIEW

File: src/types.ts [hunk 1]
```diff
@@ -0,0 +1,11 @@
+export interface ISessionStorage {
+  close(): Promise<void>;
+  set(key: string, value: any, expires_sec: number): Promise<void>;
+  tryGet(key: string): Promise<any | null>;
```
Comment (aikoven) [inline src/types.ts]: Why not just `get`?

Comment (aikoven) [reply src/types.ts]: Also don't really like the `any` here and below. Can we specify actual types?

File: src/fs.ts [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+    objRaw: fData.slice(sIdx+1)
+  }
+}
+
+class FileSessionStorage implements ISessionStorage {
+  constructor(private path: string) {}
+
+  public async close(): Promise<void> {
+    return Promise.resolve();
+  }
```
Comment (aikoven) [inline src/fs.ts]: We should stop the GC timer here

File: src/fs.ts [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+            const expires_in = fData.slice(0, fData.indexOf('\n'));
+            if (Date.now() > parseInt(expires_in, 10)) {
+              await fs.rm(ff, {force: true, maxRetries: 3, retryDelay: 10});
+            }
+          }
+        }
+      }
+    }
+  } catch (e) {
+    console.warn('file storage gc error', e);
```
Comment (aikoven) [inline src/fs.ts]: It would be better to use the Verdaccio logger instead of `console.log`

File: src/index.ts [hunk 4]
```diff
@@ -9,33 +9,38 @@ import {
...
+import { NewRedisStorage } from './redis';
+import { NewFileStorage } from './fs';
+
 type OidcPluginConfig = {
   publicUrl: string;
-  redisUri: string;
   issuer: string;
   clientId: string;
   clientSecret: string;
+  scope: string;
```
Comment (aikoven) [inline src/index.ts:27]: To maintain backward compatibility, this field must be optional and we should provide the default value for it

File: src/index.ts [hunk 5]
```diff
@@ -9,33 +9,38 @@ import {
...
   usernameClaim?: string;
   rolesClaim?: string;
+
+  redisUri?: string;
+  fs_session_store_path?: string
+  fs_token_store_path?: string;
```
Comment (aikoven) [inline src/index.ts]: These should be in camelCase for consistency

Comment (aikoven) [reply src/index.ts]: Could we also have a single option instead of two?

**Review by aikoven — COMMENTED** [review]
Hey, thanks for your contribution! This is a nice addition.

Please address the issues below

File: src/fs.ts (hunk 2 above)
Comment (nikolaevn) [reply src/fs.ts]: FIXED

File: src/index.ts (hunk 4 above)
Comment (nikolaevn) [reply src/index.ts:27]: FIXED

File: src/index.ts (hunk 5 above)
Comment (nikolaevn) [reply src/index.ts]: FIXED

File: src/types.ts (hunk 1 above)
Comment (nikolaevn) [reply src/types.ts]: FIXED

File: src/fs.ts [hunk 6]
```diff
@@ -0,0 +1,191 @@
...
+      return null;
+    }
+  }
+}
+
+class FileTokenStorage implements ITokenStorage {
+  constructor(private path: string, private fnClose: () => Promise<void>) {}
+
+  public close(): Promise<void> {
+    this.fnClose();
```
Comment (aikoven) [inline src/fs.ts]: Missing `await`

File: src/index.ts [hunk 7]
```diff
@@ -52,27 +57,17 @@ export default class OidcPlugin
...
+    if (options.config.redisUri) {
+      const rs = NewRedisStorage(options.config.redisUri);
+      this.ss = rs.ss;
+      this.ts = rs.ts;
+    } else if (options.config.fsSessionStorePath && options.config.fsTokenStorePath) {
+      const rs = NewFileStorage(this.logger, options.config.fsSessionStorePath, options.config.fsTokenStorePath);
+      this.ss = rs.ss;
+      this.ts = rs.ts;
+    } else {
+      throw new Error('invalid configuration: none of [redisUri] or [fs_session_store_path, fs_token_store_path] is set');
```
Comment (aikoven) [inline src/index.ts]:
```suggestion
      throw new Error('invalid configuration: none of [redisUri] or [fsSessionStorePath, fsTokenStorePath] is set');
```

**Review by aikoven — COMMENTED** [review]
A few more findings. Please also update the branch

File: src/fs.ts (hunk 6 above)
Comment (nikolaevn) [reply src/fs.ts]: fixed

**Comment by aikoven** [comment]
Thanks! Just released `verdaccio-openid-connect@1.2.0`

**Pull request merged**